File: aitools/storage/dummy.py
from typing import Set, Iterable, Tuple, cast, Dict

from aitools.logic import LogicObject, Substitution
from aitools.storage.base import LogicObjectStorage
from aitools.storage.index import make_key, TrieIndex, AbstruseIndex, AbstruseKey, AbstruseKeyElement


class DummyLogicObjectStorage(LogicObjectStorage):
    def __init__(self):
        self._objects: Set[LogicObject] = set()

    def add(self, *objects: LogicObject):
        for obj in objects:
            self._objects.add(obj)

    def search_unifiable(self, other: LogicObject) -> Iterable[Tuple[LogicObject, Substitution]]:
        for obj in self._objects:
            unifier = Substitution.unify(obj, other)
            if unifier is not None:
                yield obj, unifier

    def __len__(self):
        return len(self._objects)


class DummyTrieIndex(TrieIndex):
    def __init__(self):
        self.subindices: Dict[AbstruseKeyElement, DummyTrieIndex] = dict()
        self.objects: Set[DummyAbstruseIndex] = set()

    def make_node(self):
        return DummyTrieIndex()

    def _get_or_create_subindex(self, key_element) -> TrieIndex:
        if key_element not in self.subindices:
            self.subindices[key_element] = self.make_node()
        subindex: TrieIndex = self.subindices[key_element]
        return subindex

    def _get_all_subindices(self):
        return self.subindices.values()

    def _get_all_keys_and_subindices(self):
        return self.subindices.items()

    def _get_subindex_by_key_element(self, key_element: AbstruseKeyElement) -> TrieIndex:
        return self.subindices.get(key_element)

    def _maybe_store_object(self, obj):
        if obj not in self.objects:
            self.objects.add(obj)

    def _get_all_objects(self):
        return iter(self.objects)


class DummyAbstruseIndex(AbstruseIndex):

    def __init__(self, *, level=0):
        super().__init__(level=level)
        self.objects = set()
        self._subindex_tree = DummyTrieIndex()

    def make_node(self, *, new_level):
        return DummyAbstruseIndex(level=new_level)

    def _get_all_objects(self):
        return self.objects

    def _maybe_store_object(self, obj):
        if obj not in self.objects:
            self.objects.add(obj)

    @property
    def subindex_tree(self):
        return self._subindex_tree

class DummyIndexedLogicObjectStorage(LogicObjectStorage):

    def __init__(self):
        self._objects: DummyAbstruseIndex[LogicObject] = DummyAbstruseIndex()

    def add(self, *objects: LogicObject):
        for obj in objects:
            key: AbstruseKey[LogicObject] = make_key(obj)
            self._objects.add(key=key, obj=obj)

    def search_unifiable(self, other: LogicObject) -> Iterable[Tuple[LogicObject, Substitution]]:
        key: AbstruseKey[LogicObject] = make_key(other)
        for obj in self._objects.retrieve(key):
            unifier = Substitution.unify(obj, other)
            if unifier is not None:
                yield obj, unifier

    def __len__(self):
        return sum(1 for _ in self._objects.retrieve([[None]]))

File: AgainIsThisAGoodIdea/Storage/Implementations/Dummy.fs
module AITools.Storage.Implementations.Dummy

open AITools.Storage.Index
open System.Collections.Generic
open AITools.Storage.Base
open AITools.Logic.Core
open AITools.Logic.Unification

type DummyExpressionStorage() =
    inherit ExpressionStorage()

    let objects = HashSet()

    override _.Add(expressions: seq<Expression>): unit = 
        for expr in expressions do
            objects.Add(expr) |> ignore

    override _.SearchUnifiable(expression: Expression): seq<Expression * Substitution> = seq {
            for expr in objects do
                let unifier = Substitution.Unify(expr, expression)
                match unifier with
                | Some subst -> yield expr, subst
                | None -> ()
        }

    override _.Size with get () = objects.Count

    override _.Dispose () = ()

type DummyTrieIndex<'keyItem, 'item when 'keyItem: equality>() =
    inherit TrieIndex<'keyItem, 'item>()

    let mutable subindices: Dictionary<KeyElement<'keyItem>, TrieIndex<'keyItem, 'item>> = Dictionary()
    let objects = HashSet()

    override _.MaybeStoreObject(object) = objects.Add(object) |> ignore

    override _.GetOrCreateSubindex(keyElement) =
        if not (subindices.ContainsKey(keyElement))
        then subindices.Add(keyElement, DummyTrieIndex<_, _>())

        subindices.[keyElement]


    override _.Objects = seq objects

    override _.Subindices = subindices |> Seq.map (fun kvp -> kvp.Value)

    override _.KeysAndSubindices =
        subindices
        |> Seq.map (fun kvp -> (kvp.Key, kvp.Value))

    override _.GetSubindexByKeyElement(keyElement) = 
        if subindices.ContainsKey(keyElement) 
        then Some <| subindices.[keyElement]
        else None

type DummyAbstruseIndex<'keyItem, 'item when 'keyItem: equality>() =
    inherit AbstruseIndex<'keyItem, 'item, DummyAbstruseIndex<'keyItem, 'item>>()

    let objects = HashSet()

    let subindexTree = DummyTrieIndex<_, _>()

    override _.MakeNode() = DummyAbstruseIndex()
    override _.Objects = seq objects
    override _.MaybeStoreObject(object) = objects.Add(object) |> ignore
    override _.SubindexTree = upcast subindexTree

type DummyIndexedExpressionStorage() =
    inherit ExpressionStorage()

    let objects = DummyAbstruseIndex()

    override _.Add(expressions) = 
        for expr in expressions do
            let key = makeKey expr
            objects.Add(key, expr)

    override _.SearchUnifiable(expression) = seq {
            let key = makeKey(expression)
            for expr in objects.Retrieve(key) do
                let unifier = Substitution.Unify(expr, expression)
                match unifier with
                | Some subst -> yield expr, subst
                | None -> ()
        }

    override _.Size with get () = 
        Key.Wildcard 
        |> objects.Retrieve 
        |> Seq.length

    override _.Dispose () = ()